trino/auth.py

```diff
@@ -156,9 +160,112 @@ def __eq__(self, other):
-    print(auth_url)
+class RedirectHandler(metaclass=abc.ABCMeta):
+    """
+    Abstract class for Oauth redirect handlers.
+    """
+
+    @abc.abstractmethod
+    def __call__(self, url: str) -> None:
+        raise NotImplementedError()
+
+
+class ConsoleRedirectHandler(RedirectHandler):
+    """
+    Handler for OAuth redirections to log to console
+    """
+
+    def __call__(self, url: str) -> None:
+        print("Open the following URL in browser for the external authentication:")
+        print(url)
+
+
+class WebBrowserRedirectHandler(RedirectHandler):
+    """
+    Handler for OAuth redirections  to open in web browser
+    """
+
+    def __call__(self, url: str) -> None:
+        webbrowser.open_new(url)
+
+
+class CompositeRedirectHandler(RedirectHandler):
+    """
+    Composite handler for redirect handlers
+    """
+
+    def __init__(self, handlers: List[Callable[[str], None]]):
+        self.handlers = handlers
+
+    def __call__(self, url: str):
+        for handler in self.handlers:
+            handler(url)
+
+
+class _OAuth2TokenCache(metaclass=abc.ABCMeta):
+    """
+    Abstract class for Oauth token cache
+    """
+
+    @abc.abstractmethod
+    def get_token_from_cache(self, host: str) -> Optional[str]:
+        pass
+
+    @abc.abstractmethod
+    def store_token_to_cache(self, host: str, token: str) -> None:
+        pass
+
+
+class _OAuth2TokenInMemoryCache(_OAuth2TokenCache):
+    """
+    In memory Token Cache implementation
+    """
+
+    def __init__(self):
+        self._cache = {}
+
+    def get_token_from_cache(self, host: str) -> Optional[str]:
+        return self._cache.get(host)
+
+    def store_token_to_cache(self, host: str, token: str) -> None:
+        self._cache[host] = token
+
+
+class _OAuth2KeyRingTokenCache(_OAuth2TokenInMemoryCache):
+    """
+    Keyring Token Cache implementation
+    """
+
+    def __init__(self):
+        super().__init__()
+        try:
+            self._keyring = importlib.import_module("keyring")
+        except ImportError:
+            self._keyring = None
+            logger.info("keyring module not found. OAuth2 token will not be stored in keyring.")
+
+    def is_keyring_available(self) -> bool:
+        return self._keyring is not None
+
+    def get_token_from_cache(self, host: str) -> Optional[str]:
+        try:
+            return self._keyring.get_password(host, "token")
+        except self._keyring.errors.NoKeyringError:
+            pass
+
+        # keyring not available, so we fallback to the in-memory cache
+        return super().get_token_from_cache(host)
+
+    def store_token_to_cache(self, host: str, token: str) -> None:
+        try:
+            # keyring is installed, so we can store the token for reuse within multiple threads
+            self._keyring.set_password(host, "token", token)
+        except self._keyring.errors.NoKeyringError:
+            logger.info("Although keyring module is installed no backend has been detected, "
```

Comment:
Since we have this handling it should be safe to make "keyring" a default dependency instead of an "extra". WDYT?

Reply:
In most cases you don't have to install any additional packages (eg recent MacOS or Windows). Only older MacOS or Linux require installing additional packages. Also from a security perspective it may be better to let the user opt in on this behaviour. 

Now we could indeed install it as a default dependency and make this a configuration parameter. Using a configuration parameter is definitely more explicit. We could then also don't ignore this exception, and forward the exception to the user. Now actually in those cases we just do simple INFO log and rely on the standard caching. That may not be so transparent.

something like:

```
conn = connect(
    user="<username>",
    auth=OAuth2Authentication(useKeyring=True),
    http_scheme="https",
    ...
)
```

Reply:
Since it depends on things working on the OS side I think it's better to have it an "extra" then since it's then a concious choice by the user.

We can revisit using parameter to control this in future as per feedback after using this.

Reply:
I think in any case, maybe we should probably not ignore/log.info the `NoKeyRingError` and just rethrow it with a `Trino.exceptions.NotSupportedException`. I think it will be more explicit instead of silently using standard caching. Do you agree?

Reply:
Yes, if the keyring module is available but for some reason we are not using it then it should result in error shown to user instead of kind-of misleading them by using in-memory caching.